### PR TITLE
Added benchmarks to measure penalties created by boxing and dynamic dispatch

### PR DESCRIPTION
**Pull request by DarkDimius**
Created a benchmarks set that shows how big are performance penalties for boxing and dynamic dispatch for fold operation of Array[Int]

### CODE REVIEW

File: src/test/scala/scala/collection/parallel/ParArraySnippets.scala [hunk 1]
```diff
@@ -38,6 +38,42 @@ trait ParArraySnippets {
     sum
   }
 
+  def noboxing(a: Array[Int])(op:(Int, Int)=>Int) = {
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/ParArraySnippets.scala]:
```
def noboxing(a: Array[Int])(op: (Int, Int) => Int) = {
```

lets follow the conventions when it comes to spacing.

File: src/test/scala/scala/collection/parallel/ParArraySnippets.scala [hunk 2]
```diff
@@ -38,6 +38,42 @@ trait ParArraySnippets {
...
+    val until = a.length
+    var sum = a(0)
+    while (i < until) {
+      sum = op(sum, a(i))
+      i += 1
+    }
+    sum
+  }
+  
+  def boxingSpec[@specialized T](a: Array[T])(op:(T, T)=>T) = {
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/ParArraySnippets.scala]: same thing here.

File: src/test/scala/scala/collection/parallel/scalameter/Generators.scala [hunk 3]
```diff
@@ -15,6 +15,11 @@ trait Generators {
   def sizes(from: Int) = Gen.enumeration("size")(from, from * 3, from * 5)
   def ranges(from: Int) = for (size <- sizes(from)) yield 0 until size
   def arrays(from: Int) = for (size <- sizes(from)) yield (0 until size).toArray
+  def arraysBoxed(from: Int) = for (size <- sizes(from)) yield {
+    val r = new Array[java.lang.Integer](size)
+    (0 until size).foreach{x=>r(x) = x:java.lang.Integer}
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/Generators.scala]:
```
{ x => r(x) = x: java.lang.Integer }
```

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 4]
```diff
@@ -41,6 +41,36 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
 
   performance of "Par[Array]" config(opts: _*) in {
 
+    measure method "Boxing" config(opts: _*) in {
+      using(arrays(large)) curve ("Sequential") in {x=> noboxing(x)(_+_)}
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: same thing here

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 5]
```diff
@@ -123,3 +153,37 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
...
 
 
+class Helpers extends Serializable{
+
+  def sum1(source:Array[Int])(f: (Int, Int) => Int): Int = {
+    var idx = 1
+    var sum = source(0)
+    val limit1 = source.size/2
+    val limit2 = source.size
+    while(idx < limit1) { //intensionaly increasing method size, to make sure that method doesn't inline entirely
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: Same thing here, space after `//`

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 6]
```diff
@@ -123,3 +153,37 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
...
+    }
+    sum
+  }
+
+  def sum2(source:Array[Int]): Int = {
+    var idx = 1
+    var sum = source(0)
+    val limit1 = source.size/2
+    val limit2 = source.size
+    while(idx < limit1) { //intensionaly increasing method size, to make sure that method doesn't inline entirely
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: Here too.

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 7]
```diff
@@ -41,6 +41,36 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
 
   performance of "Par[Array]" config(opts: _*) in {
 
+    measure method "Boxing" config(opts: _*) in {
+      using(arrays(large)) curve ("Sequential") in {x=> noboxing(x)(_+_)}
+      using(arrays(large)) curve ("SequentialBoxedOp") in {x=> boxing(x)(_+_) }
+      using(arrays(large)) curve ("SequentialSpecialized") in {x=> boxingSpec(x)(_+_) }
+      using(arraysBoxed(small)) curve ("SequentialBoxedOpBoxedData") in {x=> boxingSpec(x)(_+_) }
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: Will this one really be specialized? You're calling `boxingSpec[@spec T]` with `T = java.lang.Integer`, so `T` is not primitive `Int`, and the non-specialized version of the `boxingSpec` should in the end be called, iiuc.
But, maybe I'm misunderstanding what's the intent of this test.

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 8]
```diff
@@ -41,6 +41,36 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
...
+      import helpers._
+      
+      var j = 0
+      var methodId = 0
+
+      using(arrays(large)) curve ("Dynamic") in { arr =>
+        val method = methodId % 5
+        methodId = methodId + 1
+
+        val inc:Int = if(method == 1) sum1(arr)(_ + _ + 3)
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]:
```
val inc: Int = if (method == 1) sum1(arr)(_ + _ + 3)
```

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 9]
```diff
@@ -41,6 +41,36 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
...
+        else if(method == 2) sum1(arr)(_ + _ + 1)
+        else if(method == 3) sum1(arr)(_ + _ + 5)
+        else if(method == 4) sum1(arr)(_ + _ + 2)
+        else  sum1(arr)(_ + _ + 4)
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: Same here.

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala [hunk 10]
```diff
@@ -41,6 +41,36 @@ class ParArrayBench extends PerformanceTest.Regression with Serializable with Pa
...
+        j = j + inc
```
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala:62]: IIUC, the idea here was to fool the JVM and prevent inlining `sum1`?

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala (hunk 7 above)
Comment (DarkDimius) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: Yes, And I wanted to create test that shows that working on boxed data(Array[Integer]) is worse than on array of primitives, and that specialization doesn't help here.

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala (hunk 10 above)
Comment (DarkDimius) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala:62]: I'm fooling it twice, method sum1 is enlarged by hand by duplicating loop, and I'm passing it multiple Funtion2 objects so JVM selects not to inline it.
I finally minimized this test yesterday from few hundred lines, I'll try to minimize it even more in future, but for now I just want the benchmakrs to appear on reports

File: src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala (hunk 7 above)
Comment (axel22) [inline src/test/scala/scala/collection/parallel/scalameter/ParArrayBench.scala]: Ok, understood.

**Comment by axel22** [comment]
LGTM!

**Pull request merged**